### PR TITLE
/tg/ Effects

### PR DESCRIPTION
**Pull request by Rohesie**
I have no idea why I've done this.
It started with fixing the `New()` proc for foam, and it ended in porting the code for /tg/ Effects, with a lot of stuff along that was needed for compatibility.

The fire effect is specific to this codebase, seems to have been added by Phil once long ago and since then forgotten. It sets things on fire and does nothing more. Tweaked it a little so it's no longer invisible, and it loops a single time instead of four.

EMP grenades and reaction work, same for the fire extinguisher's water, which use this system now.

Disabled the jetpacks, which were broken, so they are simply decorative. But this would make it easier to make them functional again.

There's support for adding features such as attack animations, or attack sounds (a few icons and sounds added).

Also changed the welding sounds to match /tg/'s.

I can atomize it if needed, but for the most part this is a refactor that changes little to our codebase, mostly touching things that don't really work currently.

### CODE REVIEW

File: code/game/objects/effects/effect_system/effects_explosion.dm [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+/datum/effect_system/explosion/set_up(loca)
+	if(isturf(loca))
+		location = loca
+	else
+		location = get_turf(loca)
+
+/datum/effect_system/explosion/start()
+	new/obj/effect/explosion( location )
+	var/datum/effect_system/expl_particles/P = new/datum/effect_system/expl_particles()
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_explosion.dm]:
```suggestion
	var/datum/effect_system/expl_particles/P = new
```

File: code/game/objects/effects/effect_system/effects_explosion.dm [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+/datum/effect_system/explosion
+
+/datum/effect_system/explosion/set_up(loca)
+	if(isturf(loca))
+		location = loca
+	else
+		location = get_turf(loca)
+
+/datum/effect_system/explosion/start()
+	new/obj/effect/explosion( location )
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_explosion.dm]:
```suggestion
	new /obj/effect/explosion(location)
```

File: code/game/objects/effects/effect_system/effects_explosion.dm [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	P.start()
+
+/datum/effect_system/explosion/smoke
+
+/datum/effect_system/explosion/smoke/proc/create_smoke()
+	var/datum/effect_system/smoke_spread/S = new
+	S.set_up(2, location)
+	S.start()
+/datum/effect_system/explosion/smoke/start()
+	..()
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_explosion.dm]:
```suggestion
	. = ..()
```

File: code/game/objects/effects/effect_system/effects_fire.dm [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+	icon = 'icons/effects/fire.dmi'
+	icon_state = "f3"
+	var/life = 0.5 SECONDS
+	mouse_opacity = 0
+
+/obj/effect/particle_effect/fire/Initialize()
+	. = ..()
+	if(!isturf(loc))
+		qdel(src)
+		return
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_fire.dm]: return the correct qdel hint instead

File: code/game/objects/effects/effect_system/effects_foam.dm [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+	if(lifetime % reagent_divisor)
+		reagents.reaction(T, VAPOR, fraction)
+
+	if(--amount < 0)
+		return
+	spread_foam()
+
+/obj/effect/particle_effect/foam/proc/foam_mob(mob/living/L)
+	if(lifetime<1)
+		return 0
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_foam.dm]:
```suggestion
		return FALSE
```

File: code/game/objects/effects/effect_system/effects_foam.dm [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+	if(!istype(L))
+		return 0
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_foam.dm]:
```suggestion
		return FALSE
```

File: code/game/objects/effects/effect_system/effects_foam.dm [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+	var/fraction = 1/initial(reagent_divisor)
+	if(lifetime % reagent_divisor)
+		reagents.reaction(L, VAPOR, fraction)
+	lifetime--
+	return 1
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_foam.dm]:
```suggestion
	return TRUE
```

File: code/game/objects/effects/effect_system/effects_foam.dm [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+		var/obj/effect/particle_effect/foam/foundfoam = locate() in T //Don't spread foam where there's already foam!
+		if(foundfoam)
+			continue
+
+		if(is_type_in_typecache(T, blacklisted_turfs))
+			continue
+
+		for(var/mob/living/L in T)
+			foam_mob(L)
+		var/obj/effect/particle_effect/foam/F = new src.type(T)
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_foam.dm]:
```suggestion
		var/obj/effect/particle_effect/foam/F = new(T)
```

File: code/game/objects/effects/effect_system/effects_foam.dm [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+
+/datum/effect_system/foam_spread/metal
+	effect_type = /obj/effect/particle_effect/foam/metal
+
+
+/datum/effect_system/foam_spread/long
+	effect_type = /obj/effect/particle_effect/foam/long_life
+
+/datum/effect_system/foam_spread/New()
+	..()
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_foam.dm]:
```suggestion
	. = ..()
```

File: code/game/objects/effects/effect_system/effects_foam.dm [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+	chemholder = new /obj()
+	var/datum/reagents/R = new/datum/reagents(1000)
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_foam.dm]:
```suggestion
	var/datum/reagents/R = new /datum/reagents(1000)
```

File: code/game/objects/effects/effect_system/effects_foam.dm [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+	chemholder.reagents = R
+	R.my_atom = chemholder
+
+/datum/effect_system/foam_spread/Destroy()
+	qdel(chemholder)
+	chemholder = null
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_foam.dm]: QDEL_NULL

File: code/game/objects/effects/effect_system/effects_other.dm [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+	qdel_in_time = 20
+
+/datum/effect_system/trail_follow/proc/set_dir(obj/effect/particle_effect/ion_trails/I)
+	I.setDir(holder.dir)
+
+//Reagent-based explosion effect
+
+/datum/effect_system/reagents_explosion
+	var/amount 						// TNT equivalent
+	var/flashing = 0			// does explosion creates flash effect?
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_other.dm]:
```suggestion
	var/flashing = FALSE			// does explosion creates flash effect?
```

File: code/game/objects/effects/effect_system/effects_other.dm [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+	var/flashing_factor = 0		// factor of how powerful the flash effect relatively to the explosion
+	var/explosion_message = 1				//whether we show a message to mobs.
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_other.dm]:
```suggestion
	var/explosion_message = TRUE				//whether we show a message to mobs.
```

File: code/game/objects/effects/effect_system/effects_other.dm [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+/datum/effect_system/reagents_explosion/set_up(amt, loca, flash = 0, flash_fact = 0, message = 1)
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_other.dm]:
```suggestion
/datum/effect_system/reagents_explosion/set_up(amt, loca, flash = FALSE, flash_fact = 0, message = TRUE)
```

File: code/game/objects/effects/effect_system/effects_other.dm [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+
+	flashing = flash
+	flashing_factor = flash_fact
+
+/datum/effect_system/reagents_explosion/start()
+	if(explosion_message)
+		location.visible_message("<span class='danger'>The solution violently explodes!</span>", \
+								"<span class='italics'>You hear an explosion!</span>")
+	if (amount < 1)
+		var/datum/effect_system/spark_spread/s = new /datum/effect_system/spark_spread
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_other.dm]:
```suggestion
		var/datum/effect_system/spark_spread/s = new
```

File: code/game/objects/effects/effect_system/effects_water.dm [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+	. = ..()
+	QDEL_IN(src, 7 SECONDS)
+
+/obj/effect/particle_effect/water/Move(turf/newloc)
+	if (--src.life < 1)
+		qdel(src)
+		return FALSE
+	if(newloc.density)
+		return FALSE
+	.=..()
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_water.dm]:
```suggestion
	return ..()
```

File: code/game/objects/effects/effect_system/effects_water.dm [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+	var/life = 15
+	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+
+
+/obj/effect/particle_effect/water/Initialize()
+	. = ..()
+	QDEL_IN(src, 7 SECONDS)
+
+/obj/effect/particle_effect/water/Move(turf/newloc)
+	if (--src.life < 1)
```
Comment (DominikPanic) [inline code/game/objects/effects/effect_system/effects_water.dm]:
```suggestion
	if (--life < 1)
```

File: code/game/objects/effects/effects.dm [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+	obj_flags = NONE
+
+
+/obj/effect/Initialize()
+	. = ..()
+	GLOB.effect_list += src
+
+
+/obj/effect/Destroy()
+	. = ..()
```
Comment (DominikPanic) [inline code/game/objects/effects/effects.dm]: This should most likely be reversed and also return ..()

File: code/modules/reagents/chemistry/recipes/pyrotechnics.dm [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+	id = "nitroglycerin"
+	results = list("nitroglycerin" = 2)
+	required_reagents = list("glycerol" = 1, "facid" = 1, "sacid" = 1)
+	strengthdiv = 2
+
+/datum/chemical_reaction/reagent_explosion/nitroglycerin/on_reaction(datum/reagents/holder, created_volume)
+	if(holder.has_reagent("stabilizing_agent"))
+		return
+	holder.remove_reagent("nitroglycerin", created_volume*2)
+	..()
```
Comment (DominikPanic) [inline code/modules/reagents/chemistry/recipes/pyrotechnics.dm]:
```suggestion
	return ..()
```

File: code/modules/reagents/chemistry/recipes/pyrotechnics.dm [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+	id = "blackpowder_explosion"
+	required_reagents = list("blackpowder" = 1)
+	required_temp = 474
+	strengthdiv = 6
+	modifier = 1
+	mix_message = "<span class='boldannounce'>Sparks start flying around the black powder!</span>"
+
+/datum/chemical_reaction/reagent_explosion/blackpowder_explosion/on_reaction(datum/reagents/holder, created_volume)
+	sleep(rand(50,100))
+	..()
```
Comment (DominikPanic) [inline code/modules/reagents/chemistry/recipes/pyrotechnics.dm]:
```suggestion
	return ..()
```

File: code/modules/reagents/chemistry/recipes/pyrotechnics.dm [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+	holder.remove_reagent("smoke_powder", created_volume*3)
+	var/smoke_radius = round(sqrt(created_volume * 1.5), 1)
+	var/location = get_turf(holder.my_atom)
+	var/datum/effect_system/smoke_spread/chem/S = new
+	S.attach(location)
+	playsound(location, 'sound/effects/smoke.ogg', 50, 1, -3)
+	if(S)
+		S.set_up(holder, smoke_radius, location, 0)
+		S.start()
+	if(holder && holder.my_atom)
```
Comment (DominikPanic) [inline code/modules/reagents/chemistry/recipes/pyrotechnics.dm]:
```suggestion
	if(holder?.my_atom)
```

File: code/modules/reagents/chemistry/recipes/pyrotechnics.dm [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
+
+/datum/chemical_reaction/reagent_explosion/nitroglycerin
+	name = "Nitroglycerin"
+	id = "nitroglycerin"
+	results = list("nitroglycerin" = 2)
+	required_reagents = list("glycerol" = 1, "facid" = 1, "sacid" = 1)
+	strengthdiv = 2
+
+/datum/chemical_reaction/reagent_explosion/nitroglycerin/on_reaction(datum/reagents/holder, created_volume)
+	if(holder.has_reagent("stabilizing_agent"))
```
Comment (Ghommie) [inline code/modules/reagents/chemistry/recipes/pyrotechnics.dm]: Stabilizing Agent is not present yet on our reagents module, but you should be able to port it here without problem.
Guess I'll have to update the interaction between heating tools and beakers for the temperature stuff to work now that recipes with temperature requirement are going in

File: code/game/machinery/computer/buildandrepair.dm [hunk 23]
```diff
@@ -28,7 +28,7 @@
 					to_chat(user, "[WT] must be on to complete this task.")
 					return
 					
-				playsound(loc, 'sound/items/Welder.ogg', 25, 1)
+				playsound(loc, 'sound/items/welder.ogg', 25, 1)
```
Comment (spookydonut) [inline code/game/machinery/computer/buildandrepair.dm]: not sure why you renamed this, the file is still called Welder/Welder2.ogg, if you were trying to rename this file lowercase it didnt work.

**Comment by spookydonut** [comment]
dont TM, will break sounds

File: code/game/atoms_movable.dm [hunk 24]
```diff
@@ -469,6 +469,67 @@
...
+	if(visual_effect_icon)
+		I = image('icons/effects/effects.dmi', A, visual_effect_icon, A.layer + 0.1)
+	else if(used_item)
+		I = image(icon = used_item, loc = A, layer = A.layer + 0.1)
+		I.plane = GAME_PLANE
+
+		// Scale the icon.
+		I.transform *= 0.75
+		// The icon should not rotate.
+		I.appearance_flags = APPEARANCE_UI_IGNORE_ALPHA
```
Comment (psykzz) [inline code/game/atoms_movable.dm]: How does IGNORE_ALPHA help with rotation?

File: code/game/atoms_movable.dm [hunk 25]
```diff
@@ -469,6 +469,67 @@
...
+		else if(direction & WEST)
+			I.pixel_x = 16
+
+		if(!direction) // Attacked self?!
+			I.pixel_z = 16
+
+	if(!I)
+		return
+
+	flick_overlay(I, GLOB.clients, 5) // 5 ticks/half a second
```
Comment (psykzz) [inline code/game/atoms_movable.dm]:
```suggestion
	flick_overlay(I, GLOB.clients, 0.5 SECONDS)
```

File: code/game/atoms_movable.dm (hunk 24 above)
Comment (Rohesie) [reply code/game/atoms_movable.dm]: This is straight out from /tg/ and I had not really analyzed it, but this is a define for more flags.
Would you like to suggest a change? This is more your alley than mine.

File: code/game/machinery/computer/buildandrepair.dm (hunk 23 above)
Comment (Rohesie) [reply code/game/machinery/computer/buildandrepair.dm]: Fixed.

**Comment by spookydonut** [comment]
DNM pending #1733

File: code/modules/reagents/chemistry/recipes/pyrotechnics.dm (hunk 22 above)
Comment (DominikPanic) [reply code/modules/reagents/chemistry/recipes/pyrotechnics.dm]: This still needs to be answered.

**Comment by Rohesie** [comment]
That PR required is merged.
Might want to test it live, because there's only so much I can test locally here.

File: code/modules/reagents/chemistry/recipes/pyrotechnics.dm (hunk 22 above)
Comment (Rohesie) [reply code/modules/reagents/chemistry/recipes/pyrotechnics.dm]: This is not a question, really. Or was there a suggestion implied?

**Comment by DominikPanic** [comment]
Still had some issues, removing testing labels as discussed in #dev-general

**Pull request closed without merging**